design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,206 @@
+
+Table funtions would also be useful for other cases such as splitting a string into words, or any
+case where you need to go from a single row to zero or more, for example, unaggregating data or
+duplicating rows.
+
+Here's an example of exploding an array from a column into multiple rows:
+
+We have a stream containing a batch of sensor readings from a particular sensor. For further
+stream processing we want to split them out into separate messages:
+
+Given:
+
+```
+{sensor_id:12345 readings: [23, 56, 3, 76, 75]}
+```
+
+We want to get:
+
+```
+{sensor_id:12345 reading: 23}
+{sensor_id:12345 reading: 56}
+{sensor_id:12345 reading: 3}
+{sensor_id:12345 reading: 76}
+{sensor_id:12345 reading: 75}
+```
+
+This could be accomplished by the following CSAS:
+
+```
+CREATE STREAM exploded_stream AS
+    SELECT sensor_id, EXPLODE(readings) AS reading FROM batched_readings;
+```
+
+Note that any scalar values in the select will be duplicated for each row in the output.
+
+
+## What is in scope
+
+We would like to implement
+
+* Table functions on the SELECT clause only
+* Multiple table functions on the SELECT clause
+* A set of built-in table functions such as EXPLODE (etc)
+* An ability for users to provide their own User Defined Table Functions (UDTFs) using annotations
+in a similar way to how UDFs and UDAFs are currently supported.
+
+## What is not in scope
+
+* Table functions on the FROM clause. We could allow table functions to take the place of other streams
+or tables in the FROM clause. We do not think there is much value in doing this as the main driver
+for this KLIP is in supporting `explode` functionality which is more simply implemented on the
+SELECT clause.
+* Table functions that output more than one column. Some databases support the flattening of an array
+ of N STRUCT types into N rows of M columns, where there is one column for each field of the struct.
+ Instead we propose that flattening an array of N structs will result in N rows with a single column
+ of type STRUCT. The fields of the struct can then be extracted if necessary using a further CSAS
+ which pulls out the struct fields.
+
+## Value/Return
+
+This functionality will make it easier for users to make end-end streaming apps using KSQL
+alone without having to have custom stages using Kafka Streams.
+
+We believe its a common case for users to have array or map like data structures as values
+in messages that they wish to explode. This new feature would allow them to do that in a pure KSQL
+app thus allowing them to create richer KSQL applications.
+
+## Public APIs
+
+There will be no changes to the KSQL language or any existing public APIs or configuration files.
+
+We will introduce new annotations to allow users to mark their UDTFs, in a very similar way to
+UDAFs.
+
+## Design
+
+### General plumbing
+
+We will need the following general plumbing to support table functions:
+
+* In a similar way to how UDAFs are currently supported we will need to create a 
```

Comment:
unless I'm missing something, UDTFs take scalar inputs right and don't need an undo/merge? In that case, can we model them after UDFs instead of after UDAFs? I think the model is easier to understand (there is one `@Udf` method for each UDF and no `Factory`).